mlir/lib/Dialect/MIOpen/Transforms/CloneKernels.cpp

```diff
@@ -28,50 +28,65 @@
 #include "llvm/ADT/SmallVector.h"
 
 using namespace mlir;
+using namespace llvm;
 
 namespace {
 
 struct MIOpenCloneKernelsPass
     : public MIOpenCloneKernelsPassBase<MIOpenCloneKernelsPass> {
+  MIOpenCloneKernelsPass(ArrayRef<StringRef> _chips)
+      : chips{_chips.begin(), _chips.end()} {}
+
   void runOnOperation() override {
+    const char *kernel = "kernel.module";
     ModuleOp mod = getOperation();
-    llvm::StringRef modName = miopen::MIOpenDialect::kKernelModuleName;
-    if (mod.getName() == modName)
+    if (mod->hasAttr(kernel))
       return;
 
-    auto *ctx = &getContext();
-    auto gpuMod = mod.lookupSymbol<ModuleOp>(modName);
-    if (!gpuMod) {
-      OpBuilder b(ctx);
-      // create a GPUModuleOp in case the GPU module specified does not exist.
-      gpuMod = b.create<ModuleOp>(mod.getLoc(), modName);
+    for (auto chip : chips) {
+      std::string modName = ("__kernel_" + chip).str();
 
-      // add the GPUModuleOp into the symbol table.
-      SymbolTable symbolTable(mod);
-      symbolTable.insert(gpuMod);
-    }
-    SymbolTable symbolTable(gpuMod);
+      auto *ctx = &getContext();
+      auto kernelMod = mod.lookupSymbol<ModuleOp>(modName);
+      if (!kernelMod) {
+        OpBuilder b(ctx);
+        // create a KERNEL ModuleOp in case the KERNEL module specified does not
+        // exist.
+        kernelMod = b.create<ModuleOp>(mod.getLoc(), StringRef(modName));
+        kernelMod->setAttr(kernel, b.getUnitAttr());
+        kernelMod->setAttr("kernel.chip", b.getStringAttr(chip));
 
-    for (auto func : mod.getOps<func::FuncOp>()) {
-      if (func->hasAttr("kernel")) {
-        // clone the func
-        auto gpuFunc = func.clone();
-        gpuFunc->setAttr("original_func", SymbolRefAttr::get(func));
+        // add the KERNELModuleOp into the symbol table.
+        SymbolTable symbolTable(mod);
+        symbolTable.insert(kernelMod);
+      }
+      SymbolTable symbolTable(kernelMod);
 
-        // add the GPUModuleOp into the symbol table.
-        symbolTable.insert(gpuFunc);
+      for (auto func : mod.getOps<func::FuncOp>()) {
+        if (func->hasAttr("kernel")) {
+          // clone the func
+          auto kernelFunc = func.clone();
+          kernelFunc->setAttr("original_func", SymbolRefAttr::get(func));
 
-        // TODO: also find all calls and import callee
+          // add the KERNELModuleOp into the symbol table.
+          symbolTable.insert(kernelFunc);
+
+          // TODO: also find all calls and import callee
+        }
       }
     }
   }
+
+private:
+  SmallVector<StringRef, 4> chips;
```

Comment:
The storage of `StringRef` requires knowledge of the lifetimes of the underlying strings, which I don't think you have. This'll probably cause dangling pointers. `SmallVector<SmallString<8>>` or such instead?